### PR TITLE
Hardware Class Interface

### PR DESCRIPTION
**Pull request by blhough**
Let's agree on an initial interface for the Hardware class.

### CODE REVIEW

**Comment by ids1024** [comment]
Do we want to have an interface that both the simulator and the real hardware class can implement?

**Comment by blhough** [comment]
Previously what we did was use #ifdefs so when in simulation `move` would just update the position instantly and do nothing with `Hardware` and `checkWall` would look up in a generated maze instead.

But we can go all out and try to simulate the actual movement.

If we decide to do the minimum simulation then `simulate` would only need a mouse object and a reference to the maze.

If we decide to go all out we will need to be more creative with how to separate the simulation from the actual code.

I'm open to any ideas.

File: Hardware.h [hunk 1]
```diff
@@ -1,14 +1,41 @@
...
+    // Returns the corresponding result after calling `checkWall`.
+    bool isWall(Relation relation);
+
+    void calibrateMotors();
+    void calibrateRangeFinders();
 
   private:
+    RangeFinder rangeFinders[NUM_RANGE_FINDERS];
+    Motor leftMotor;
+    Motor rightMotor;
```
Comment (klanmiko) [inline Hardware.h:46]: We should create variables to define the rotation angle, distance moved, based on encoder values.

File: Hardware.h [hunk 2]
```diff
@@ -1,14 +1,41 @@
...
+    // static const unsigned RANGE_FINDER_FRONT_PIN = 18
+
+    static const unsigned NUM_RANGE_FINDERS = 3;
+
     Hardware();
-    ~Hardware();
+
+    void moveForward(unsigned mm);
+    void rotate(int deg);
+
```
Comment (klanmiko) [inline Hardware.h]: Return a bool value to check whether the robot encountered an obstacle.

Also how would we allow the Mouse class to know that the robot has moved entirely from one cell to another?

**Review by klanmiko — COMMENTED** [review]
Need to allow the Mouse class to know the current coordinate and direction

File: Hardware.h (hunk 2 above)
Comment (blhough) [reply Hardware.h]: The movement functions would be blocking. When the movement is complete control will return to the calling object.

File: Hardware.h (hunk 1 above)
Comment (blhough) [reply Hardware.h:46]: Right I was thinking those would go in the motor class, but we could also have it in this class.

https://docs.google.com/spreadsheets/d/1TA4xhWMv8Iiz3RvajErwgmy2hwtNH2dCPY6-RLwRgxc/edit#gid=0

**Pull request merged**